src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonTypeInfoResolverTests.JsonPropertyInfo.cs

```diff
@@ -1178,5 +1178,195 @@ private class MyClassCustomConverterFactory : JsonConverterFactory
+        {
+            var options = new JsonSerializerOptions
+            {
+                TypeInfoResolver = new DefaultJsonTypeInfoResolver()
+                {
+                    Modifiers =
+                    {
+                        jti =>
+                        {
+                            if (jti.Type == typeof(ClassWithExtensionDataAttribute))
+                            {
+                                Assert.Equal(3, jti.Properties.Count);
+                                Assert.True(jti.Properties[2].IsExtensionData);
+                                jti.Properties.RemoveAt(2);
+                            }
+                        }
+                    }
+                }
+            };
+
+            var value = new ClassWithExtensionDataAttribute
+            {
+                Value1 = 1,
+                Value2 = 2,
+                ExtensionData = new Dictionary<string, object>
+                {
+                    ["Value3"] = 3
+                }
+            };
+
+            string json = JsonSerializer.Serialize(value, options);
+            JsonTestHelper.AssertJsonEqual("""{"Value1":1,"Value2":2}""", json);
+
+            value = JsonSerializer.Deserialize<ClassWithExtensionDataAttribute>("""{"unrecognizedValue":42}""", options);
+            Assert.Null(value.ExtensionData);
+        }
+
+        [Theory]
+        [InlineData(typeof(IDictionary<string, object>))]
+        [InlineData(typeof(IDictionary<string, JsonElement>))]
+        [InlineData(typeof(Dictionary<string, object>))]
+        [InlineData(typeof(Dictionary<string, JsonElement>))]
+        [InlineData(typeof(ConcurrentDictionary<string, JsonElement>))]
+        [InlineData(typeof(JsonObject))]
+        public static void EnablingExtensionData_SupportedPropertyType(Type type)
+        {
+            var jti = JsonTypeInfo.CreateJsonTypeInfo(typeof(ClassWithExtensionDataAttribute), new());
+            var jpi = jti.CreateJsonPropertyInfo(type, "ExtensionData");
+
+            Assert.False(jpi.IsExtensionData);
+            jpi.IsExtensionData = true;
+            Assert.True(jpi.IsExtensionData);
+        }
+
+        [Theory]
+        [InlineData(typeof(int))]
+        [InlineData(typeof(string))]
+        [InlineData(typeof(object))]
+        [InlineData(typeof(ClassWithExtensionDataAttribute))]
+        [InlineData(typeof(List<int>))]
+        [InlineData(typeof(IDictionary<string, string>))]
+        [InlineData(typeof(Dictionary<string, int>))]
+        public static void EnablingExtensionData_UnsupportedPropertyType_ThrowsInvalidOperationException(Type type)
+        {
+            var jti = JsonTypeInfo.CreateJsonTypeInfo(typeof(ClassWithExtensionDataAttribute), new());
+            var jpi = jti.CreateJsonPropertyInfo(type, "ExtensionData");
+
+            Assert.False(jpi.IsExtensionData);
+            Assert.Throws<InvalidOperationException>(() => jpi.IsExtensionData = true);
```

Comment:
so if user incorrectly places the attribute on something which is not meant to be extension data it's impossible to correct that error? should we be doing validation in the configure to allow to fix the incorrect user setting?

Reply:
I settled for throwing behavior right now because pushing it to `Configure` necessitated substantial refactoring of the `CacheMember` logic. I attempted doing that but ended up breaking a bunch of member visibility tests in subtle ways.  We can revisit the behavior and move all validation to `Configure` in vNext (at the same time we fix all the `JsonIgnoreAttribute` bugs)

FWIW we currently also throw early when we encounter other invalid configuration, e.g. when we have conflicting property names.

Reply:
Why substantial refactoring? we only need to not throw on IsExtensionData and just set it to true based on attribute presence. We move throwing into configure (IsExtensionData + invalid type)

Reply:
It's not just validating that the type is right, there's also the concern of validating duplicate annotations. In order to make this possible, we need to stop building the property cache early in the case of reflection metadata. But, like I said this requires refactoring of the member resolution which I attempted. We need to do this eventually, but it should not regress member visibility behavior and it should not regress performance in the default use (i.e. reflection-based metadata without custom resolvers should not necessitate building the property cache twice).